File: Python/Medium/2327_Number_of_People_Aware_of_a_Secret/solution.py
class Solution:
    def peopleAwareOfSecret(self, n: int, delay: int, forget: int) -> int:
        return Solution.person(1, 1, n, delay, forget, {})
        
    def person(start, currentDay, n, delay, forget, memo):
        MOD = 10**9 + 7
        
        key = (start, currentDay)
        if key in memo:
            return memo[key]

        if currentDay > n:
            if n - start < forget:
                return 1
            else:
                return 0 

        count = 0
        if start + delay <= currentDay and currentDay - start < forget:
            count += Solution.person(currentDay, currentDay + 1, n, delay, forget, memo) % MOD

        count += Solution.person(start, currentDay + 1, n, delay, forget, memo) % MOD
        
        memo[key] = count
        return count

File: Golang/Medium/2327_Number_of_People_Aware_of_a_Secret/solution.go
package numberofpeopleawareofasecret

type state struct {
	start      int
	currentDay int
}

func peopleAwareOfSecret(n int, delay int, forget int) int {
	return person(1, 1, n, delay, forget, make(map[state]int))
}

func person(start int, currentDay int, n int, delay int, forget int, memo map[state]int) int {
	const mod = 1_000_000_007
	key := state{start, currentDay}
	if v, exist := memo[key]; exist {
		return v
	}

	if currentDay > n {
		if n-start < forget {
			return 1
		} else {
			return 0
		}
	}

	count := 0
	if start+delay <= currentDay && currentDay-start < forget {
		count += person(currentDay, currentDay+1, n, delay, forget, memo) % mod
	}

	count += person(start, currentDay+1, n, delay, forget, memo) % mod

	memo[key] = count
	return count
}
